File: Java/src/com/bh/no0020_Valid_Parentheses.java
package com.bh;

import java.util.Stack;

/**
 * LeetCode 20: Valid Parentheses
 * 
 * Given a string s containing just the characters '(', ')', '{', '}', '[' and ']',
 * determine if the input string is valid.
 * 
 * A string is valid if:
 * 1. Open brackets must be closed by the same type of brackets.
 * 2. Open brackets must be closed in the correct order.
 * 3. Every close bracket has a corresponding open bracket of the same type.
 */
public class no0020_Valid_Parentheses {

    static class Solution {
        public boolean isValid(String s) {
            Stack<Character> stack = new Stack<>();
            for (char c : s.toCharArray()) {
                if (c == '(' || c == '[' || c == '{') {
                    stack.push(c);
                } else {
                    if (stack.isEmpty()) return false;
                    char top = stack.pop();
                    if (c == ')' && top != '(') return false;
                    if (c == ']' && top != '[') return false;
                    if (c == '}' && top != '{') return false;
                }
            }
            return stack.isEmpty();
        }
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        // Test cases
        String[] testCases = {"()", "()[]{}", "(]", "([])", "([)]", "{[]}", "(", ")", "", "(((())))"};
        boolean[] expected = {true, true, false, true, false, true, false, false, true, true};
        for (int i = 0; i < testCases.length; i++) {
            boolean result = solution.isValid(testCases[i]);
            System.out.println("Input: '" + testCases[i] + "' | Output: " + result + " | Expected: " + expected[i]);
        }
    }
}

File: CPlusPlus/no0020_Valid_Parentheses.cpp
//
// Created by binghuan on 2025/9/17.
//
#include <iostream>
#include <stack>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> st; // Stack to store opening brackets

        for (char c : s) {
            // If it's an opening bracket, push to stack
            if (c == '(' || c == '{' || c == '[') {
                st.push(c);
            }
            // If it's a closing bracket
            else if (c == ')' || c == '}' || c == ']') {
                // If stack is empty, no matching opening bracket
                if (st.empty()) {
                    return false;
                }

                char top = st.top();
                st.pop();

                // Check if the closing bracket matches the opening bracket
                if ((c == ')' && top != '(') ||
                    (c == '}' && top != '{') ||
                    (c == ']' && top != '[')) {
                    return false;
                }
            }
        }

        // If stack is empty, all brackets are properly matched
        return st.empty();
    }
};

// Test function
void testCase(const string& s, bool expected) {
    Solution solution;
    bool result = solution.isValid(s);
    cout << "Input: \"" << s << "\" -> Output: " << (result ? "true" : "false");
    cout << " (Expected: " << (expected ? "true" : "false") << ")";
    cout << " [" << (result == expected ? "PASS" : "FAIL") << "]" << endl;
}

int main() {
    cout << "=== LeetCode 20: Valid Parentheses ===" << endl;
    cout << endl;

    // Test cases from the problem
    testCase("()", true);           // Example 1
    testCase("()[]{}", true);       // Example 2
    testCase("(]", false);          // Example 3
    testCase("([])", true);         // Example 4
    testCase("([)]", false);        // Example 5

    // Additional test cases
    testCase("", true);             // Empty string
    testCase("((()))", true);       // Nested parentheses
    testCase("()[]", true);         // Multiple pairs
    testCase("(())", true);         // Nested single type
    testCase("({[]})", true);       // All types nested
    testCase("({[}])", false);      // Wrong order
    testCase("((", false);          // Only opening
    testCase("))", false);          // Only closing
    testCase("())", false);         // Extra closing
    testCase("(()", false);         // Extra opening

    cout << endl;
    cout << "Test completed!" << endl;

    return 0;
}
